productserver/main.go

```diff
@@ -0,0 +1,175 @@
+
+}
+func GetHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	err := json.NewEncoder(w).Encode(product)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func GetIDHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+	intid := 0
+	_, err := fmt.Sscan(id, &intid)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	for _, products := range product {
+		if intid > len(product) {
+			w.WriteHeader(http.StatusNotFound)
+			response := "404 status not found"
+			_, err = w.Write([]byte(response))
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+			return
+		}
+		if products.ID == intid {
+			response, err := json.Marshal(product[intid-1])
+			if err != nil {
+				log.Print("error 3")
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+			w.WriteHeader(http.StatusOK)
+
+			_, err = w.Write(response)
+		}
+	}
+}
+func PostHandler(w http.ResponseWriter, r *http.Request) {
+	if err := r.ParseForm(); err != nil {
+		fmt.Fprintf(w, "ParseForm() err: %v", err)
+		return
+	}
+	var newProduct Product
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+	}
+
+	err = json.Unmarshal(reqBody, &newProduct)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	product = append(product, newProduct)
+	w.WriteHeader(http.StatusCreated)
+
+	json.NewEncoder(w).Encode(newProduct)
+
+}
+func PutHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+	intid := 0
+	_, err := fmt.Sscan(id, &intid)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	for index, products := range product {
+		if intid > len(product) {
+			w.WriteHeader(http.StatusNotFound)
+			response := "404 status not found"
+			_, err = w.Write([]byte(response))
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+			return
+		}
+		if products.ID == intid {
+			product = append(product[:index], product[:index+1]...)
+			err = json.NewDecoder(r.Body).Decode(&product)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+			products.ID = intid
+			product = append(product, products)
+			json.NewEncoder(w).Encode(product)
+			return
+		}
+	}
+}
+func DeleteHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+	intid := 0
+	_, err := fmt.Sscan(id, &intid)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	for index, products := range product {
+		if intid > len(product) {
+			w.WriteHeader(http.StatusNotFound)
+			response := "404 status not found"
+			_, err = w.Write([]byte(response))
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+			return
+		}
+		if products.ID == intid {
+			product = append(product[:index], product[index+1:]...)
+			response := "Product Deleted"
+			_, err = w.Write([]byte(response))
```

Comment:
No need for this response. Communicate with the right response header status code.